pinot-spi/src/main/java/org/apache/pinot/spi/stream/MessageBatch.java

```diff
@@ -116,4 +116,14 @@ default StreamPartitionMsgOffset getOffsetOfNextBatch() {
   default boolean isEndOfPartitionGroup() {
     return false;
   }
+
+  /**
+   * This is useful while determining ingestion delay for a message batch. Retaining metadata for last message in
+   * a batch can enable us to estimate the ingestion delay for the batch.
+   *
+   * @return null by default.
+   */
+  default public StreamMessageMetadata getLastMessageMetadata() {
```

Comment:
nit: Add `@Nullable` annotation to this method 

Can we do better than returning `null`? maybe, by default, check the message at the last index and if it exists, return its metadata. if there are no messages, return null. Do you think this is better?

Reply:
Ack to nullable.

I think the second suggestion in this comment may lead to trouble: the last index may not be the message that provided the last metadata: it depends on errors etc and the external code already did the filtering for us so I don't want to have to redo that logic here.

Reply:
>the last index may not be the message that provided the last metadata: it depends on errors etc and the external code already did the filtering for us so I don't want to have to redo that logic here. 

The fact that last index may not be the message that provides the last metadata is not clear to me. 

Say the batch contains consumed = "[m1, m2, m3, m4]" and after filtering = "[m1 , m2, <filtered>, m4]". 
Is such a batch this possible?  If yes, what should be the return value of `getLastMessageMetadata` ? m3 ?

The javadoc says "Retaining metadata for last message in a batch can enable us to estimate the ingestion delay for the batch." . So, it is m4?

Reply:
Addressed in last commit